synapse/storage/persist_events.py

```diff
@@ -734,7 +770,105 @@ async def _get_new_state_after_events(
+        state_resolutions_during_persistence.inc()
+
+        # If the returned state matches the state group of one of the new
+        # forward extremities then we check if we are able to prune some state
+        # extremities.
+        if res.state_group and res.state_group in new_state_groups:
+
+            new_latest_event_ids = await self._prune_extremities(
+                room_id,
+                new_latest_event_ids,
+                res.state_group,
+                event_id_to_state_group,
+                events_context,
+            )
+
+        return res.state, None, new_latest_event_ids
+
+    async def _prune_extremities(
+        self,
+        room_id: str,
+        new_latest_event_ids: Set[str],
+        resolved_state_group: int,
+        event_id_to_state_group: Dict[str, int],
+        events_context: List[Tuple[EventBase, EventContext]],
+    ) -> Set[str]:
+        """See if we can prune any of the extremities after calculating the
+        resolved state.
+        """
+        potential_times_prune_extremities.inc()
+
+        # We keep all the extremities that have the same state group, and
+        # see if we can drop the others.
+        new_new_extrems = {
+            e
+            for e in new_latest_event_ids
+            if event_id_to_state_group[e] == resolved_state_group
+        }
+
+        dropped = set(new_latest_event_ids) - new_new_extrems
+
+        logger.debug("Might drop events: %s", dropped)
+
+        # We only drop events if:
+        #   1. we're not currently persisting them;
+        #   2. they're not our own events (or are dummy events); and
+        #   3. they're either:
+        #       1. over N hours old and more than N events ago (we use depth
+        #          to calculate); or
+        #       2. we are persisting an event from the same domain.
+        #
+        # The idea is that we don't want to drop events that are "legitmate"
+        # extremities (that we would want to include as prev events), only
+        # "stuck" extremities that are e.g. due to a gap in the graph.
+        #
+        # Note that we either drop all of them or none of them. If we only
+        # drop some of the events we don't know if state res would come to
+        # the same conclusion.
+
+        for ev, _ in events_context:
+            if ev.event_id in dropped:
+                logger.debug("Not dropping as in to persist list")
+                return new_latest_event_ids
+
+        dropped_events = await self.main_store.get_events(
+            dropped, allow_rejected=True, redact_behaviour=EventRedactBehaviour.AS_IS,
+        )
+
+        new_senders = {get_domain_from_id(e.sender) for e, _ in events_context}
+
+        one_day_ago = self._clock.time_msec() - 20 * 60 * 60 * 1000
+        current_depth = max(e.depth for e, _ in events_context)
+        for event in dropped_events.values():
+            if self.is_mine_id(event.sender) and event.type != "org.matrix.dummy_event":
```

Comment:
should use a constant from `EventTypes` for this, to make for easier searching. (I'm disappointed to find we don't already have one)

Reply:
This heuristic feels magical. What makes it safe to ignore dummy events?

Reply:
> This heuristic feels magical. What makes it safe to ignore dummy events?

My, perhaps flawed, thinking is that since we generally only send dummy events in rooms that the server isn't sending "proper" events into, old dummy events are likely to only point to old remote events, which we are OK with dropping.

The reason to do this is that dummy events are going to be a source of stuck extremities, so it would be good to be able to prune them.

Reply:
> old dummy events are likely to only point to old remote events, which we are OK with dropping.

the implication here is that it is *not* ok to drop forward extremities that are local events. You haven't spelt out the reason for this, but presumably it's to avoid us leaving dangling extremities across the whole federation?

the thing is, it's absolutely not a given that all a dummy event's predecessors are remote events, particularly in a situation where federation starts dropping out due to load on one side or the other (or there simply aren't enough "real" events moving around to de-extremify any local events). So dropping such a dummy event would be equivalent to dropping those local predecessors as extremities.

I wonder if we could/should strengthen the test to: "if it's a dummy event, then it's ok to drop it if either (1) all the dummy's predecessors are remote or (2) any local predecessors have other successors (that we aren't dropping)."

Reply:
> the implication here is that it is _not_ ok to drop forward extremities that are local events. You haven't spelt out the reason for this, but presumably it's to avoid us leaving dangling extremities across the whole federation?

So my reasoning to not drop local extremities is to give the server a chance to include them as prev events when sending new events, so that other servers will notice if they didn't get the previous extremities. I don't think dropping local, stale extremities will cause them to dangle elsewhere? It's just that either a) we missed a remote event that referenced it or b) we failed to send it out to the server that is generating events (if not us).

> the thing is, it's absolutely not a given that all a dummy event's predecessors are remote events, particularly in a situation where federation starts dropping out due to load on one side or the other (or there simply aren't enough "real" events moving around to de-extremify any local events). So dropping such a dummy event would be equivalent to dropping those local predecessors as extremities.

Yes, though to be a candidate for pruning the dummy event needs to either be old, or for the server to be generating events. I think the danger here is that the server ends up with many forward extremities, so new events don't reference all of the existing extremities, and so we prune extremities  even though they weren't "stale" as such.



> wonder if we could/should strengthen the test to: "if it's a dummy event, then it's ok to drop it if either (1) all the dummy's predecessors are remote or (2) any local predecessors have other successors (that we aren't dropping)."

This is actually fine to do, for some reason I didn't think that was a thing we could look at. Will add

Reply:
Actually, as a first cut I might just only check that dummy events point to remote events. That should catch the majority of stuck dummy extremities.

Reply:
I've implemented dropping of dummy events if and only if they only reference remote events or other dummy events that only reference remote events (etc).